src/sys/wasi/p2.rs

```diff
@@ -0,0 +1,675 @@
+
+#[cfg(feature = "net")]
+use {
+    crate::Registry,
+    std::{ffi::c_int, net::SocketAddr},
+};
+
+#[allow(unused_macros)]
+macro_rules! syscall {
+    ($fn: ident ( $($arg: expr),* $(,)* ) ) => {{
+        let res = unsafe { libc::$fn($($arg, )*) };
+        if res == -1 {
+            Err(std::io::Error::last_os_error())
+        } else {
+            Ok(res)
+        }
+    }};
+}
+
+cfg_net! {
+    pub(crate) mod tcp {
+        use std::io;
+        use std::os::fd::AsRawFd;
+        use std::net::{self, SocketAddr};
+        use std::ffi::c_int;
+        use super::{new_socket, socket_addr, netc as libc};
+
+        pub(crate) fn accept(listener: &net::TcpListener) -> io::Result<(net::TcpStream, SocketAddr)> {
+            let (stream, addr) = listener.accept()?;
+            stream.set_nonblocking(true)?;
+            Ok((stream, addr))
+        }
+
+        pub(crate) fn new_for_addr(address: SocketAddr) -> io::Result<c_int> {
+            let domain = match address {
+                SocketAddr::V4(_) => libc::AF_INET,
+                SocketAddr::V6(_) => libc::AF_INET6,
+            };
+            new_socket(domain, libc::SOCK_STREAM)
+        }
+
+        pub(crate) fn connect(socket: &net::TcpStream, addr: SocketAddr) -> io::Result<()> {
+            let (raw_addr, raw_addr_length) = socket_addr(&addr);
+
+            match syscall!(connect(
+                socket.as_raw_fd(),
+                raw_addr.as_ptr(),
+                raw_addr_length
+            )) {
+                Err(err) if err.raw_os_error() != Some(libc::EINPROGRESS) => Err(err),
+                _ => Ok(()),
+            }
+        }
+    }
+}
+
+#[cfg(all(debug_assertions, feature = "net"))]
+static NEXT_ID: AtomicUsize = AtomicUsize::new(1);
+
+#[derive(Debug, Copy, Clone)]
+struct Subscription {
+    token: Token,
+    interests: Option<Interest>,
+}
+
+pub(crate) struct Selector {
+    #[cfg(all(debug_assertions, feature = "net"))]
+    id: usize,
+    subscriptions: Arc<Mutex<HashMap<RawFd, Subscription>>>,
+}
+
+impl Selector {
+    pub(crate) fn new() -> io::Result<Selector> {
+        Ok(Selector {
+            #[cfg(all(debug_assertions, feature = "net"))]
+            id: NEXT_ID.fetch_add(1, Ordering::Relaxed),
+            subscriptions: Arc::new(Mutex::new(HashMap::new())),
+        })
+    }
+
+    #[cfg(all(debug_assertions, feature = "net"))]
+    pub(crate) fn id(&self) -> usize {
+        self.id
+    }
+
+    pub(crate) fn try_clone(&self) -> io::Result<Selector> {
+        Ok(Selector {
+            #[cfg(all(debug_assertions, feature = "net"))]
+            id: self.id,
+            subscriptions: self.subscriptions.clone(),
+        })
+    }
+
+    pub(crate) fn select(&self, events: &mut Events, timeout: Option<Duration>) -> io::Result<()> {
+        events.clear();
+
+        let mut subscriptions = self.subscriptions.lock().unwrap();
+
+        let mut states = Vec::new();
+        for (fd, subscription) in subscriptions.deref() {
```

Comment:
I don't think looping of all active subscriptions is a great idea. Nor is the allocation for `states`.

Why switch from a `Vec` to a `HashMap`?

Reply:
I'm happy to optimize by reducing the number of allocations if desired.  Would you prefer that I update this PR or leave that for a follow-up PR?

> Why switch from a Vec to a HashMap?

I was aiming for O(1) lookups by file descriptor in the `IoSourceState` implementation.

Reply:
> I'm happy to optimize by reducing the number of allocations if desired. Would you prefer that I update this PR or leave that for a follow-up PR?
> 
> > Why switch from a Vec to a HashMap?
> 
> I was aiming for O(1) lookups by file descriptor in the `IoSourceState` implementation.

It might be slower when taking into account hashing tho :thinking:

Reply:
> It might be slower when taking into account hashing tho 🤔

Possibly; I can switch back to using a `Vec` if preferred.